source/device_defender.c

```diff
@@ -350,6 +529,61 @@ static void s_reporting_task_fn(struct aws_task *task, void *userdata, enum aws_
                 goto cleanup;
             }
 
+            if (AWS_OP_SUCCESS != s_init_custom_metric_data(custom_metric_data, defender_task)) {
+                goto cleanup;
+            }
+
+            if (custom_metrics_len != 0) {
+                for (size_t metric_index = 0; metric_index < custom_metrics_len; ++metric_index) {
+                    aws_array_list_get_at(
+                        &defender_task->config.custom_metrics,
+                        (void *)&custom_metric_data[metric_index].metric,
+                        metric_index);
+
+                    /* TEMP checkpoint for debugging */
+                    AWS_LOGF_DEBUG(
+                        AWS_LS_IOTDEVICE_DEFENDER_TASK,
+                        "id=%p Retrieving value for custom metric %s",
+                        (void *)defender_task,
+                        aws_string_c_str(custom_metric_data[metric_index].metric->metric_name));
+
+                    int result = AWS_OP_SUCCESS;
+                    switch (custom_metric_data[metric_index].metric->type) {
+                        case DD_METRIC_NUMBER:
+                            result = custom_metric_data[metric_index].metric->supplier_fn.get_number_fn(
+                                &custom_metric_data[metric_index].data.number,
+                                custom_metric_data[metric_index].metric->userdata);
+                            break;
+                        case DD_METRIC_NUMBER_LIST:
+                            result = custom_metric_data[metric_index].metric->supplier_fn.get_number_list_fn(
+                                &custom_metric_data[metric_index].data.list,
+                                custom_metric_data[metric_index].metric->userdata);
+                            break;
+                        case DD_METRIC_STRING_LIST:
+                            result = custom_metric_data[metric_index].metric->supplier_fn.get_string_list_fn(
+                                &custom_metric_data[metric_index].data.list,
+                                custom_metric_data[metric_index].metric->userdata);
+                            break;
+                        case DD_METRIC_IP_LIST:
+                            result = custom_metric_data[metric_index].metric->supplier_fn.get_ip_list_fn(
+                                &custom_metric_data[metric_index].data.list,
+                                custom_metric_data[metric_index].metric->userdata);
+                            break;
+                        case DD_METRIC_UNKNOWN:
+                        default:
+                            /* output warning or error */
+                            continue;
+                            break;
+                    }
+
+                    if (result != AWS_OP_SUCCESS) {
+                        /* TODO: bad result */
```

Comment:
Should we havve a log line at least here until the implementation comes?

Reply:
Implemented. I decided that each instance of the custom metric data would also store the result of the retrieval, so during the report generation, it can know if the metric was *not* retrieved successfully and skip processing it. It cleans up the JSON a bit if entire lists are considered bad data, and is entirely necessary to know that an int64_t is bad without using a sentinel value.